src/api/environment.cc

```diff
@@ -67,6 +67,15 @@ MaybeLocal<Value> PrepareStackTraceCallback(Local<Context> context,
   if (env == nullptr) {
     return exception->ToString(context).FromMaybe(Local<Value>());
   }
+  // TODO(legendecas): Per-realm prepareStackTrace callback.
+  // If we are in a Realm that is not the principal Realm (e.g. ShadowRealm),
+  // skip the prepareStackTrace callback as the context's security token is
+  // likely to be different.
```

Comment:
Actually, why do we need different security tokens? Can we just set the non-principal realm's security token to be the same as the one of the principal realm, like what we do for vm contexts? IIUC security tokens in browsers are generally intended for cross-origin global proxies, which isn't really a thing for Node.js. For shadow realms, the cross-realm access is guarded with wrapped functions and callable boundaries. For Node.js realms, do we care about cross-realm object access?

Reply:
It is true that object exchange across the ShadowRealm boundary is limited with wrapped functions. However, it is still possible for host hooks like the prepareStackTrace callback here to leak objects to the principal realm's userland `Error.prepareStackTrace` override.

It doesn't hurt to set the security token to be the principal realm's. But I don't see the reason to allow the access either.

I've updated the comment to point out that this branch is intended to avoid calling the principal realm's `Error.prepareStackTrace` override instead of the security token mismatches.